beacon-chain/blockchain/service.go

```diff
@@ -222,26 +226,228 @@ func (c *ChainService) blockProcessing(processedBlock chan<- *types.Block) {
 		// or via p2p.
 		case block := <-c.incomingBlockChan:
-			if err := c.processBlock(block); err != nil {
-				log.Error(err)
-				processedBlock <- nil
+
+			// Before sending the blocks for processing we check to see if the blocks
+			// are valid to continue being processed. If the slot number in the block
+			// has already been processed by the beacon node, we throw it away. If the
+			// slot number is too high to be processed in the current slot, we store
+			// it in a cache.
+
+			beaconState, err := c.beaconDB.GetState()
+			if err != nil {
+				log.Errorf("Unable to retrieve beacon state %v", err)
+				continue
+			}
+
+			currentSlot := beaconState.Slot()
+
+			if currentSlot+1 < block.SlotNumber() {
+				c.unProcessedBlocks[block.SlotNumber()] = block
 				continue
 			}
 
-			// Push the block to trigger the fork choice rule.
-			processedBlock <- block
+			if currentSlot+1 == block.SlotNumber() {
+
+				if err := c.receiveBlock(block); err != nil {
+					log.Error(err)
+					processedBlock <- nil
+					continue
+				}
+
+				// Push the block to trigger the fork choice rule.
+				processedBlock <- block
+			} else {
+				log.Debugf(
+					"Block slot number is lower than the current slot in the beacon state %d",
+					block.SlotNumber())
+				c.sendAndDeleteCachedBlocks(currentSlot)
+			}
 		}
 	}
 }
 
-func (c *ChainService) processBlock(block *types.Block) error {
-	blockHash, err := block.Hash()
+// receiveBlock is a function that defines the operations that are preformed on
+// any block that is received from p2p layer or rpc. It checks the block to see
+// if it passes the pre-processing conditions, if it does then the per slot
+// state transition function is carried out on the block.
+// spec:
+//  def process_block(block):
+//      if not block_pre_processing_conditions(block):
+//          return False
+//
+//  	# process skipped slots
+//
+// 		while (state.slot < block.slot - 1):
+//      	state = slot_state_transition(state, block=None)
+//
+//		# process slot with block
+//		state = slot_state_transition(state, block)
+//
+//		# check state root
+//      if block.state_root == hash(state):
+//			return state
+//		else:
+//			return False  # or throw or whatever
+//
+func (c *ChainService) receiveBlock(block *types.Block) error {
+
+	blockhash, err := block.Hash()
 	if err != nil {
-		return fmt.Errorf("failed to get hash of block: %v", err)
+		return fmt.Errorf("could not hash incoming block: %v", err)
 	}
 
-	if c.enablePOWChain && !c.doesPoWBlockExist(block) {
-		return errors.New("proof-of-Work chain reference in block does not exist")
+	beaconState, err := c.beaconDB.GetState()
+	if err != nil {
+		return fmt.Errorf("failed to get beacon state: %v", err)
+	}
+
+	if block.SlotNumber() == 0 {
+		return errors.New("cannot process a genesis block: received block with slot 0")
+	}
+
+	// Save blocks with higher slot numbers in cache.
+	if !c.isBlockReadyForProcessing(block) && block.SlotNumber() > beaconState.Slot() {
```

Comment:
We are doing the `block.slot > state.slot` check [here](https://github.com/prysmaticlabs/prysm/pull/1083/files#diff-7f087f690e8ec9178851c9b6844cb529R244), do we need to check that twice?

Reply:
Also should it be `block.SlotNumber() > beaconState.Slot()+1` ?

Reply:
yeap you are right, will get it changed